src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+            var options = document?.Project?.Solution?.Options;
+            var highlightMatchingPortions = options?.GetOption(CompletionOptions.HighlightMatchingPortionsOfCompletionListItems, document.Project.Language) ?? true;
+            var showCompletionItemFilters = options?.GetOption(CompletionOptions.ShowCompletionItemFilters, document.Project.Language) ?? true;
+
+            var updatedFilters = showCompletionItemFilters
+                ? GetUpdatedFilters(initialListOfItemsToBeIncluded, data.SelectedFilters)
+                : ImmutableArray<AsyncCompletionData.CompletionFilterWithState>.Empty;
+
+            var highlightedList = GetHighlightedList(initialListOfItemsToBeIncluded, filterText, highlightMatchingPortions).ToImmutableArray();
+
+            // If this was deletion, then we control the entire behavior of deletion ourselves.
+            if (initialRoslynTriggerKind == CompletionTriggerKind.Deletion)
+            {
+                return HandleDeletionTrigger(initialListOfItemsToBeIncluded, filterText, updatedFilters, highlightedList);
+            }
+
+            Func<ImmutableArray<RoslynCompletionItem>, string, ImmutableArray<RoslynCompletionItem>> filterMethod;
+            if (completionService == null)
+            {
+                filterMethod = (items, text) => CompletionService.FilterItems(_completionHelper, items, text);
+            }
+            else
+            {
+                filterMethod = (items, text) => completionService.FilterItems(document, items, text);
+            }
+
+            return HandleNormalFiltering(
+                filterMethod,
+                filterText,
+                updatedFilters,
+                initialRoslynTriggerKind,
+                filterReason,
+                data.Trigger.Character,
+                initialListOfItemsToBeIncluded,
+                highlightedList,
+                hasSuggestedItemOptions);
+        }
+
+        private static bool IsAfterDot(ITextSnapshot snapshot, ITrackingSpan applicableToSpan)
+        {
+            var position = applicableToSpan.GetStartPoint(snapshot).Position;
+            return position > 0 && snapshot[position - 1] == '.';
+        }
+
+        private AsyncCompletionData.FilteredCompletionModel HandleNormalFiltering(
+            Func<ImmutableArray<RoslynCompletionItem>, string, ImmutableArray<RoslynCompletionItem>> filterMethod,
+            string filterText,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters,
+            CompletionTriggerKind initialRoslynTriggerKind,
+            CompletionFilterReason filterReason,
+            char typeChar,
+            List<ExtendedFilterResult> itemsInList,
+            ImmutableArray<AsyncCompletionData.CompletionItemWithHighlight> highlightedList,
+            bool hasSuggestedItemOptions)
+        {
+            // Not deletion.  Defer to the language to decide which item it thinks best
+            // matches the text typed so far.
+
```

Comment:
missing the code that bails out if we can't find a completion service.  why was that code removed?

Reply:
Checks for the completion service were moved to the root method UpdateCompletionList. However, we do not bail out if there is not service: we have to support calls to UpdateCompletionList even for items we didn't produce before.

---
In reply to: [232510223](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510223) [](ancestors = 232510223)

Reply:
See above.  I would really like to understand and discuss this new design requirement/change.  Can we please not consider this resolved until then.  Thanks!

Reply:
We had discussions with @amadeusw and @jasonmalinowski. Here is the current idea:

1. Editor decides whom to call for CompletionSource upfront.
2. Editor selects a single ItemManager based on CompletionSources chosen again upfront. This decision cannot be changed later due to asynchronous approaches.
3. Roslyn is interested in being ItemManager, i.e. providing 
Filter and Sort. Most of other providers are currently not interested in this. There are currently no known conflicts between neighbor providers who can want to be an ItemManager for the same completion scenario. Moreover, most providers do not implement ItemManager at all. Maybe this could change in the future.
4. Then, Editor may ask or may not ask Roslyn for items (CompletionSource), Roslyn may want or may not want to participate (CompletionSource.InitializeCompletion) and Roslyn may provide items or may provide not items (CompletionSource.GetCompletionContextAsync). However, if Editor decided that Roslyn is an item manager, Roslyn will be called for filtering and sorting even if there are no items provided by Roslyn.

If in the future we will see more providers who want to implement ItemManager, we will have to re-consider filtering (especially) and sorting.

Based on this, we consider the current design to be valid for current purposes. @cyrusnajmabadi and @sharwell, please let me know if you have any comments or concerns.

Resolving all current issues. Would like to get a signoff from @jasonmalinowski as well. Expect to get it on 11/19.

---
In reply to: [233701344](https://github.com/dotnet/roslyn/pull/29016#discussion_r233701344) [](ancestors = 233701344,232896712)